{"contract": "0xd5967fed03e85d1cce44cab284695b41bc675b5c", "sourceLanguage": "Solidity", "count_risk_of_false_positives": 1, "list_vulnerabilities": ["ArithmeticVulnerability", "DenialServiceVulnerability", "UncheckedLowCallsVulnerability", "ReentrancyVulnerability", "OtherVulnerability"], "rules": [{"id": "Arithmetic_4", "fullDescription": {"text": "A possible integer overflow exists in the function `transfer(address,address,address[],uint256)`.\nThe addition or multiplication may result in a value higher than the maximum representable integer."}, "name": "ArithmeticVulnerability", "shortDescription": {"text": "Integer Overflow"}}, {"id": "Reentrancy_3", "fullDescription": {"text": "This contract executes a message call to an address provided as a function argument. Generally, it is not recommended to call user-supplied addresses using Solidity's call() construct. Note that attackers might leverage reentrancy attacks to exploit race conditions or manipulate this contract's state."}, "name": "ReentrancyVulnerability", "shortDescription": {"text": "Message call to external contract"}}, {"id": "Other_10", "fullDescription": {"text": "Multiple sends exist in one transaction, try to isolate each external call into its own transaction. As external calls can fail accidentally or deliberately.\nConsecutive calls: \nCall at address: 652\nCall at address: 652\nCall at address: 652\nCall at address: 652\nCall at address: 652\n"}, "name": "OtherVulnerability", "shortDescription": {"text": "Multiple Calls"}}, {"id": "UncheckedLowCalls_3", "fullDescription": {"text": "The return value of an external call is not checked. Note that execution continue even if the called contract throws."}, "name": "UncheckedLowCallsVulnerability", "shortDescription": {"text": "Unchecked CALL return value"}}, {"id": "DenialService_3", "fullDescription": {"text": "demo.transfer has external calls inside a loop:\n\t- caddress.call(id,from,_tos[i],v) (/data/demo_analysis/source_code/0xd5967fed03e85d1cce44cab284695b41bc675b5c.sol#16-18)\n"}, "name": "DenialServiceVulnerability", "shortDescription": {"text": "calls-loop"}}, {"id": "UncheckedLowCalls_5", "fullDescription": {"text": "Low level call in demo.transfer (/data/demo_analysis/source_code/0xd5967fed03e85d1cce44cab284695b41bc675b5c.sol#11-20):\n\t-caddress.call(id,from,_tos[i],v) /data/demo_analysis/source_code/0xd5967fed03e85d1cce44cab284695b41bc675b5c.sol#16-18\n"}, "name": "UncheckedLowCallsVulnerability", "shortDescription": {"text": "low-level-calls"}}, {"id": "Arithmetic_11", "name": "ArithmeticVulnerability", "shortDescription": {"text": "Integer Overflow."}}, {"id": "DenialService_2", "name": "DenialServiceVulnerability", "shortDescription": {"text": "Callstack Depth Attack Vulnerability."}}], "listLine": [11, 16], "analysis": {"11": [{"level": "warning", "tool": ["mythril", "oyente"], "snippet": {"text": "ransfer(address from,address caddress,address[] _tos,uint v)public returns (bool){\n        require(_tos.length > 0);\n        bytes4 id=bytes4(keccak256(\"transferFrom(address,address,uint256)\"));\n        for(uint i=0;i<_tos.length;i++){\n            // <yes> <report> UNCHECKED_LL_CALLS\n            caddress.call(id,from,_tos[i],v);\n        }\n        return true;\n    }\n}"}, "fullDescription": "A possible integer overflow exists in the function `transfer(address,address,address[],uint256)`.\nThe addition or multiplication may result in a value higher than the maximum representable integer.", "name": "ArithmeticVulnerability"}, {"level": "warning", "tool": ["slither"], "fullDescription": "demo.transfer has external calls inside a loop:\n\t- caddress.call(id,from,_tos[i],v) (/data/demo_analysis/source_code/0xd5967fed03e85d1cce44cab284695b41bc675b5c.sol#16-18)\n", "name": "DenialServiceVulnerability"}, {"level": "note", "tool": ["slither"], "fullDescription": "Low level call in demo.transfer (/data/demo_analysis/source_code/0xd5967fed03e85d1cce44cab284695b41bc675b5c.sol#11-20):\n\t-caddress.call(id,from,_tos[i],v) /data/demo_analysis/source_code/0xd5967fed03e85d1cce44cab284695b41bc675b5c.sol#16-18\n", "name": "UncheckedLowCallsVulnerability"}], "16": [{"level": "warning", "tool": ["mythril"], "snippet": {"text": "d,from,_tos[i],v);\n        }\n   "}, "fullDescription": "This contract executes a message call to an address provided as a function argument. Generally, it is not recommended to call user-supplied addresses using Solidity's call() construct. Note that attackers might leverage reentrancy attacks to exploit race conditions or manipulate this contract's state.", "name": "ReentrancyVulnerability"}, {"level": "warning", "tool": ["mythril"], "fullDescription": "Multiple sends exist in one transaction, try to isolate each external call into its own transaction. As external calls can fail accidentally or deliberately.\nConsecutive calls: \nCall at address: 652\nCall at address: 652\nCall at address: 652\nCall at address: 652\nCall at address: 652\n", "name": "OtherVulnerability", "flag": "risk of false positives"}, {"level": "note", "tool": ["mythril", "slither"], "fullDescription": "The return value of an external call is not checked. Note that execution continue even if the called contract throws.", "name": "UncheckedLowCallsVulnerability"}, {"level": "warning", "tool": ["slither", "oyente"], "fullDescription": "demo.transfer has external calls inside a loop:\n\t- caddress.call(id,from,_tos[i],v) (/data/demo_analysis/source_code/0xd5967fed03e85d1cce44cab284695b41bc675b5c.sol#16-18)\n", "name": "DenialServiceVulnerability"}]}, "count_vulnerabilities": {"ArithmeticVulnerability": 1, "DenialServiceVulnerability": 2, "UncheckedLowCallsVulnerability": 2, "ReentrancyVulnerability": 1, "OtherVulnerability": 1}, "count_leve_vulnerabilities": {"warning": 5, "error": 0, "note": 2, "none": 0}, "false_positives": 0}